sql/core/src/main/scala/org/apache/spark/sql/expressions/UserDefinedFunction.scala

```diff
@@ -85,8 +94,9 @@ case class UserDefinedFunction protected[sql] (
    * @since 2.3.0
    */
   def withName(name: String): this.type = {
-    this._nameOption = Option(name)
-    this
+    val udf = copyAll()
+    udf._nameOption = Option(name)
```

Comment:
I know your intention here (you probably mean we should not update values even in `var` variables) though, is it okay that the code below has four times object allocation in the worst case? I'm a bit worried about this point;

```
val udf = UserDefinedFunction(func, dataType, inputTypes).withName(name).withNullability(nullable)
val withDeterminism = if (!deterministic) udf.nonDeterministic() else udf
val withDistinctLike = if (distinctLike) withDeterminism.withDistinctLike() else withDeterminism
```